api/ai/new/geg_algorithm/geg_algorithm.py

```diff
@@ -0,0 +1,352 @@
+                self.remove_edge(project_id, other_project_id)
+                self.add_edge(other_project_id, project_id)
+
+            # In case utility decreases, will start envy with more and some will stop envying
+            if updated_utility < self.additive_utilities[other_project_id]:
+                self.add_edge(project_id, other_project_id)
+                self.remove_edge(other_project_id, project_id)
+
+            # In case utility is equal to other, noone will envy
+            if updated_utility == self.additive_utilities[other_project_id]:
+                self.remove_edge(project_id, other_project_id)
+                self.remove_edge(other_project_id, project_id)
+
+    def get_all_cycles(self) -> List[List[int]]:
+        visited = set([])
+        cycles = []
+
+        for start_node in self.graph:
+            self._dfs(start_node, visited, [])
+
+        return cycles
+
+    def _dfs(self, node: int, visited: Set[int], path: List[int]) -> List[int]:
+        if node in path:
+            return path[path.index(node) :]
+
+        if node not in visited:
+            visited.add(node)
+            path.append(node)
+            for neighbor in self.get_envy_targets(node):
+                self._dfs(neighbor, visited, path)
+            path.pop()
+            visited.remove(node)
+
+
+def requirement_met_by_student(
+    requirement: ProjectRequirement, student: Student
+) -> bool:
+    is_met = False
+    for value in student.attributes.get(requirement.attribute):
+        if requirement.operator == RequirementOperator.LESS_THAN:
+            is_met |= value < requirement.value
+        elif requirement.operator == RequirementOperator.MORE_THAN:
+            is_met |= value > requirement.value
+        else:  # default case is 'exactly'
+            is_met |= value == requirement.value
+    return is_met
+
+
+class GEGAlgorithm(Algorithm):
+    """
+    Generalized Envy Graph Algorithm
+    """
+
+    envy_graph: EnvyGraph
+    trace_dictionary: Dict[int, Student]
+
+    algorithm_run_time: float
+
+    def __init__(
+        self,
+        algorithm_options: GeneralizedEnvyGraphAlgorithmOptions,
+        team_generation_options: TeamGenerationOptions,
+        projects: List[Project],
+    ):
+        super().__init__(algorithm_options, team_generation_options)
+
+        self.allocation: Dict[int, List[int]] = {}
+        self.utilities: Dict[Tuple[int, int], int] = {}
+        self.projects: List[Project] = projects
+
+    def _randomize_utilities(
+        self, students: List[Student]
+    ) -> Dict[Tuple[int, int], int]:
+        utilities: Dict[Tuple[int, int], int] = {}
+
+        for project in self.projects:
+            for student in students:
+                utilities[(project.id, student.id)] = random.randint(-5, 5)
+
+        return utilities
+
+    def _get_N_plus(self, projects: List[Project], student: Student) -> List[Project]:
```

Comment:
I'm guessing you named these functions/variables in a way that aligns with the formulas in the papers you read, but remember this is code that someone else will read, and some of these function/variable names make kinda no sense.

I see a few options:
1. Keep wacky function names, add better comments
2. Make better function names, add comments to explain what they correlate to in the papers
3. ??? be creative